helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixAdmin.java

```diff
@@ -2176,36 +2192,40 @@ public ZKHelixAdmin build() {
     }
   }
 
-  private Map<String, InstanceConfig> findTimeoutOfflineInstances(String clusterName,
-      long offlineDuration) {
-    Map<String, InstanceConfig> instanceConfigMap = new HashMap<>();
+  private Set<String> findTimeoutOfflineInstances(String clusterName, long offlineDuration) {
     // in case there is no customized timeout value, use the one defined in cluster config
     if (offlineDuration == ClusterConfig.OFFLINE_DURATION_FOR_PURGE_NOT_SET) {
       offlineDuration =
           _configAccessor.getClusterConfig(clusterName).getOfflineDurationForPurge();
       if (offlineDuration == ClusterConfig.OFFLINE_DURATION_FOR_PURGE_NOT_SET) {
-        return instanceConfigMap;
+        return Collections.emptySet();
       }
     }
 
     HelixDataAccessor accessor =
         new ZKHelixDataAccessor(clusterName, new ZkBaseDataAccessor<ZNRecord>(_zkClient));
     PropertyKey.Builder keyBuilder = accessor.keyBuilder();
-    instanceConfigMap = accessor.getChildValuesMap(keyBuilder.instanceConfigs(), true);
+    List<String> instanceConfigNames = accessor.getChildNames(keyBuilder.instanceConfigs());
+    List<String> instancePathNames = accessor.getChildNames(keyBuilder.instances());
     List<String> liveNodes = accessor.getChildNames(keyBuilder.liveInstances());
-    instanceConfigMap.keySet().removeAll(liveNodes);
 
-    Set<String> toRemoveInstances = new HashSet<>();
-    for (String instanceName : instanceConfigMap.keySet()) {
+    Set<String> timeoutOfflineInstanceNames = new HashSet<>();
+    timeoutOfflineInstanceNames.addAll(instanceConfigNames);
+    timeoutOfflineInstanceNames.addAll(instancePathNames);
+    liveNodes.forEach(timeoutOfflineInstanceNames::remove);
+    for (String instanceName : timeoutOfflineInstanceNames) {
       ParticipantHistory participantHistory =
           accessor.getProperty(keyBuilder.participantHistory(instanceName));
-      long lastOfflineTime = participantHistory.getLastOfflineTime();
-      if (lastOfflineTime == ClusterConfig.OFFLINE_DURATION_FOR_PURGE_NOT_SET
-          || System.currentTimeMillis() - lastOfflineTime < offlineDuration) {
-        toRemoveInstances.add(instanceName);
+
+      // If participant history is null, a race condition happened and should be cleaned up
+      // Otherwise, if the participant has not been offline for more than the duration, no clean up
+      if (participantHistory != null
```

Comment:
You're correct that the newly created nodes may be impacted. This is a conundrum: the whole point of this change is to clean up leftover ZNodes caused by race conditions (messages sent to deleted ZNodes) because otherwise those leftovers can never be purged, however, fundamentally these ZNodes cannot be distinguished from the partial creation of new instances. 

The reason why I did it this way is because the pervious implementation warns to not use this endpoint while instances are joining. The assumption is the same for this implementation, then we also don't need to worry about new instances (although it has changed from "while no old instances are rejoining" to "while no instances are joining, old or new"). 

If we really want to keep the behavior as before, we can check for two cases:
1. Existing InstanceConfig but no History -> partial creation of new instances, don't purge;
2. Non-existing InstanceConfig and no History -> race condition, purge.
But do you think that's necessary, given the previously mentioned assumption of using this endpoint?

Reply:
I guess as long as we keep the assumption that no new node will join the cluster during purging, the logic should be fine. However, it is less ideal since there is no mechanism to enforce it. For now, I think additional checks, which introduce a lot of complexity, might not be enough to cover all the use cases. Let's just keep it simple.
But I would suggest adding a warning log if an instance that is being deleted might be a new node.

Reply:
It didn't sit well with me so I changed the logic - now it should exclude new instances. The logic is not complex in the end.